pyclam/criterion.py

```diff
@@ -130,38 +130,103 @@ def __call__(self, root: Cluster) -> Set[Cluster]:
             return {cluster for cluster in manifold.layers[self.depth].clusters}
 
 
+def _percentile_selection(manifold: Manifold, predicted_auc: Dict[Cluster, float]) -> Set[Cluster]:
+    """ Selects clusters for graph when their predicted auc is greater than the p-th percentile of the predicted auc of their descendents.
+    """
+    graph: Set[Cluster] = set()
+    clusters: List[Cluster] = [child for child in manifold.root.children]
+    while clusters:
+        new_clusters: List[Cluster] = list()
+        for cluster in clusters:
+            if cluster.children:
+                subtree_auc = [predicted_auc[c] for c in cluster.descendents]
+                if predicted_auc[cluster] >= np.percentile(subtree_auc, q=75):
+                    graph.add(cluster)
+                else:
+                    new_clusters.extend(cluster.children)
+            else:
+                graph.add(cluster)
+        clusters = new_clusters
+    return graph
+
+
+def _ranked_selection(manifold: Manifold, predicted_auc: Dict[Cluster, float]) -> Set[Cluster]:
+    """ Rank-orders clusters by predicted_auc and peel off the top to build the graph.
+    """
+    rankings: Dict[Cluster, float] = {cluster: 0 for cluster in predicted_auc.keys()}
+    for leaf in manifold.layers[-1].clusters:
+        # sort ancestors by predicted auc in non-increasing order, break ties by smaller depth
+        ancestors = list(sorted(manifold.ancestry(leaf), key=lambda c: (predicted_auc[c], c), reverse=True))
+        # add index of cluster in sorted list of ancestors to rankings
+        for i, cluster in enumerate(ancestors):
+            rankings[cluster] += (i * cluster.cardinality)
+    rankings.pop(manifold.select(''))  # remove root from rankings because ratios are undefined
+
+    # TODO: misses some points/clusters. breaks graph invariant
```

Comment:
is this TODO still true?

Reply:
I haven't checked since the Graph refactor. There was an insidious, hidden bug in the old version of Graph that showed up in all sorts of places. I hope this is fixed now. I'll check in the other PR